transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,636 @@
+            reference
+                .new_connections
+                .unbounded_send(Ok(ListenerEvent::NewAddress(address)))
+                .expect("we have a reference to the peer, so this will not fail; qed");
+        }
+        let endpoint = EndpointRef { reference, channel };
+        let join_handle = spawn(endpoint.clone());
+        Ok((Self(endpoint), join_handle))
+    }
+}
+
+fn create_muxer(
+    endpoint: ConnectionEndpoint,
+    connection: Connection,
+    handle: ConnectionHandle,
+    inner: &mut EndpointInner,
+) -> Upgrade {
+    super::connection::ConnectionDriver::spawn(endpoint, connection, handle, |weak| {
+        inner.muxers.insert(handle, weak);
+    })
+}
+
+impl EndpointData {
+    fn accept_muxer(
+        self: &Arc<Self>,
+        connection: Connection,
+        handle: ConnectionHandle,
+        inner: &mut EndpointInner,
+        channel: Channel,
+    ) {
+        let connection_endpoint = ConnectionEndpoint(EndpointRef {
+            reference: self.clone(),
+            channel,
+        });
+        let upgrade = create_muxer(connection_endpoint, connection, handle, &mut *inner);
+        if self
+            .new_connections
+            .unbounded_send(Ok(ListenerEvent::Upgrade {
+                upgrade,
+                local_addr: self.address.clone(),
+                remote_addr: self.address.clone(),
+            }))
+            .is_err()
+        {
+            inner.inner.accept();
+            inner.inner.reject_new_connections();
+        }
+    }
+}
+
+impl Future for EndpointRef {
+    type Output = Result<(), Error>;
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<(), Error>> {
+        let Self { reference, channel } = self.get_mut();
+        let mut inner = reference.inner.lock();
+        trace!("driving events");
+        inner.drive_events(cx);
+        trace!("driving incoming packets");
+        match inner.drive_receive(&reference.socket, cx)? {
+            Poll::Pending => {}
+            Poll::Ready((handle, connection)) => {
+                trace!("have a new connection");
+                reference.accept_muxer(connection, handle, &mut *inner, channel.clone());
+                trace!("connection accepted");
+            }
+        }
+        match inner.poll_transmit_pending(&reference.socket, cx)? {
+            Poll::Pending | Poll::Ready(()) => {
+                let refcount = Arc::strong_count(&reference);
+                trace!("Strong reference count is {}", refcount);
+                if refcount == 1 {
+                    debug!("All connections have closed. Closing QUIC endpoint driver.");
+                    drop(inner);
+                    Poll::Ready(Ok(()))
+                } else {
+                    Poll::Pending
+                }
+            }
+        }
+    }
+}
+
+/// A QUIC listener
+#[derive(Debug)]
+pub struct Listener {
+    reference: Arc<EndpointData>,
+    /// MUST COME AFTER `reference`!
```

Comment:
I don't get it

Reply:
Rust runs destructors in declaration order.  The `Arc` needs to be dropped first, to avoid a deadlock.

Reply:
Specifically, the destructor of `Channel` wakes up the `Endpoint`, so that it can see that the reference count has hit 1 and exit.